parsl/monitoring/db_local.py

```diff
@@ -0,0 +1,218 @@
+                            # if this is the first sight of a task, create a task_status_table to hold this task's updates
+                            if 'task_status' not in self.meta.tables.keys():
+                                task_status_table = create_task_status_table(self.meta)
+                                self.meta.create_all(con)
+                                con.execute(task_status_table.insert().values(**{k: v for k, v in info.items() if k in task_status_table.c}))
+                            # if this status table already exists, just insert the update
+                            else:
+                                task_status_table = self.meta.tables['task_status']
+                                con.execute(task_status_table.insert().values(**{k: v for k, v in info.items() if k in task_status_table.c}))
+                            return
+
+                        if 'psutil_process_pid' in info.keys():
+                            # if this is a task resource update then handle that, if the resource table DNE then create it
+                            if 'task_resources' not in self.meta.tables.keys():
+                                task_resource_table = create_task_resource_table(self.meta)
+                                self.meta.create_all(con)
+                                con.execute(task_resource_table.insert().values(**{k: v for k, v in info.items() if k in task_resource_table.c}))
+                            # if this resource table already exists, just insert the update
+                            else:
+                                task_resource_table = self.meta.tables['task_resources']
+                                con.execute(task_resource_table.insert().values(**{k: v for k, v in info.items() if k in task_resource_table.c}))
+                            return
+
+            except Exception as e:
+                print(str(e))
+                failed = True
+                time.sleep(5)
+            if not failed:
+                return
+
+
+class RemoteHandler(logging.Handler):
+    """ Handler used to pass a log to the logging server for it to be written to the database.
+    This handler is added to a logger that is returned by get_db_logger when requested. """
+    def __init__(self, web_app_host, web_app_port, request_timeout=40, retries=3, on_fail_sleep_duration=5):
+        """ Set up various behavior of the attempt to connect to the logging server. Not currently configurable. """
+        logging.Handler.__init__(self)
+        self.addr = web_app_host + ':' + str(web_app_port)
+        self.request_timeout = request_timeout
+        self.retries = retries
+        self.on_fail_sleep_duration = on_fail_sleep_duration
+
+    def emit(self, record):
+        """ Open up an HTTP connection to the logging server. This is currently blocking.
+        It should not matter for the resource monitors but could hold up the DFK logs if this connection/sending takes a lot of time. """
```

Comment:
does this blocking only hold up logging, or will it hold up workflow progression in general?

Reply:
While the logging server is handling a log, all other logs are queued holding up logging. When a log from the DFK (for example a task status update) is submitted, it blocks while it lines up in the server's queue and will block until the request is handled or it has failed too many times. To avoid this blocking, I could try to implement the asynchronous HTTP request which I am currently unfamiliar with. In action, I'm not sure how worrisome or impactful this blocking is but I think it is important to have considered/mentioned.